### ISSUE TITLE
Weighted Training by Target and Data

### ISSUE THREAD

**Issue by SoodabehGhaffari** [chemprop/chemprop#1031]
Hello,
I have a very imbalanced binary dataset and I would like to add weights to data points. The weights are not only different between the two classes, but also data points that belong to the positive class should have different weights. According to Chemprop tutorial, weights should be negative. I have some questions:
1. Is it common to set zero for a negative class and one for a positive class?
2. I would like to know if there is any paper which has used weights in their Chemprop models?
3. Do you have any suggestions about specifying weights in training?

I appreciate your help.
Best Regards
Soo

**Comment by shihchengli** [comment]
Hi @SoodabehGhaffari, as far as I know, the weights should be positive. Could you point out where you found information suggesting that the weights should be negative? Also, for imbalanced classes, you can consider using the `--class-balance` flag, which we recently implemented in the CLI (https://github.com/chemprop/chemprop/pull/1011), but you can already access it via the API.

> Is it common to set zero for a negative class and one for a positive class?

Yes, the true class should be labeled as 1, and the false class should be labeled as 0. However, I also notice that some people would use -1 and 1. In Chemprop, please use 0 and 1.

> I would like to know if there is any paper which has used weights in their Chemprop models?

I don't know of many papers that have used data weights in Chemprop. I quickly searched and found one. They used data weights to prevent repetitive molecules from contributing to large weight updates during the model training process.
https://www.nature.com/articles/s41597-022-01154-3

> Do you have any suggestions about specifying weights in training?

It really depends on the purpose of your model training. I cannot provide a general guideline for how to assign data weights, as it varies based on your specific goals. Some people want their model to perform equally well across the entire range of the predicted property, so they weight the data according to the distribution of values (similar to what you want to do). In some cases, certain data points may be noisier or less representative of the overall distribution, making it reasonable to assign them lower weights.

**Comment by SoodabehGhaffari** [comment]
Thank you for the quick response and suggestions!
There was a typo in my question. According to the Chemprop tutorial, the
weights should be *non-negative.*

I used -class-balance flag in my current model, however, the Precision
Recall AUC is below 20% while the ROC AUC is around 75%. We tested the top
predictions, none showed activity. As a result,  I need to retrain the
model. To do so, I am planning to remove the --class-balance flag and use
weights. What do you think about this idea? Do you have any
other suggestions to increase the accuracy of the model?

I appreciate your help.
Best Regards
Soo

**Comment by shihchengli** [comment]
You could try it, as using class balance would reduce the amount of data for model training. FWIW, it is also important to understand the reference for your PR AUC.
https://stats.stackexchange.com/questions/424515/pr-auc-50-with-roc-auc-90-model-good-or-bad

**Comment by SoodabehGhaffari** [comment]
Thank you very much for your help.
Best,
Soodabeh

**Comment by shihchengli** [comment]
I think you're referring to task weights. For a regression task, the loss function, MSE, would be the average MSE across every targets. If you don't normalize the targets, your model would bias towards learning one of them.

**Comment by SoodabehGhaffari** [comment]
Thanks for your reply. Let me clarify my question:
I have a binary dataset which is very imbalanced. I would like to set
weights for each data point to not only distinguish between positive and
negative classes, but also assign different weights to different positive
classes. In this situation, can  I use the weights that are not normalized?
Thanks you for your help
Best regards,
Soo

**Comment by shihchengli** [comment]
I don't think you need to normalize the data weights. Please point out to me where you see the statement of "Values are normalized to make the average weight equal 1. " so I can better understand your question.

**Comment by SoodabehGhaffari** [comment]
Here is the paragraph:
Using the --data_weights_path argument followed by a path to a data file
containing weights will allow each individual datapoint in the training
data to be given different weight in parameter updates. Formatting of this
file is similar to provided features CSV files: they should contain only a
single column with one header row and a numerical value in each row that
corresponds to the order of datapoints provided with --data_path. Data
weights should not be provided for validation or test sets if they are
provided through the arguments --separate_test_path or --separate_val_path.
Provided weights must be non-negative. *Values are normalized to make the
average weight equal 1*. Data weights are not used with the validation set
for the determination of early stopping or in evaluation of the test set.

Here is the link for the paragraph above:
https://github.com/chemprop/chemprop/blob/v1.7.1/README.md#weighted-training-by-target-and-data

Please let me know when we need to normalize based on the paragraph above
and how it helps. I appreciate your help.

Best Regards
Soo

**Comment by shihchengli** [comment]
I feel it might not be necessary to normalize those weights. TWIW, normalizing the data weights can ensure they have an average value of one so the overall scale of the loss remains the same regardless of the weights.
